Use Xilinx IP in `callisto` for synthesizability

martijnbastiaan opened the pull request:
This still needs some work, mainly in the form of cautionary comments. Hence, this PR is still a draft.

martijnbastiaan commented:
Even though it's a draft, feel free to review @vmchale !

martijnbastiaan commented:
This is done, however, I want to merge it after https://github.com/bittide/bittide-hardware/pull/161 so I can move the synthesizable `callisto` to `bittide` (and prevent merge conflicts).

lmbollen reviewed (approved):
LGTM modulo optimizations and some small comments.

I'm still missing a test we can run on CI that verifies the correct functioning of the control algorithm. Given that we're currently prioritizing a running clock demo, I'll approve this PR so it does not hold up other PRs and I will make an issue to implement a test for CI.

vmchale commented on elastic-buffer-sim/src/Bittide/ClockControl/Strategies/Callisto.hs:
```diff
@@ -2,81 +2,130 @@
+  (HiddenClockResetEnable dom, KnownNat n, 1 <= n) =>
+  -- | Target data count. See 'targetDataCount'.
+  DataCount ->
+  -- | Provide an update every /n/ cycles
```
You can use TeX in haddock!

https://haskell-haddock.readthedocs.io/en/latest/markup.html

martijnbastiaan replied:
I'm going to move forward with the merge, given that this isn't critical. Please make a papercut PR if you feel I should have done a different mark up style :).

vmchale commented on elastic-buffer-sim/src/Bittide/ClockControl/Strategies/Callisto.hs:
```diff
@@ -2,81 +2,130 @@
+    let
+      -- These bit coerces are fairly dangerous / unchecked.
+      --
+      -- TODO: Defend datatypes (why they don't overflow), or propagate errors up.
```
Should this be done before merging?

martijnbastiaan replied:
No, 32 bit should be plenty for our demos. I'm trying to speedrun getting things working on the board.

martijnbastiaan commented on elastic-buffer-sim/src/Bittide/ClockControl/Strategies/Callisto.hs:
```diff
@@ -2,81 +2,130 @@
+initState :: ControlSt
+initState = ControlSt 0 0 NoChange
+
+-- | A counter that starts at a given value, counts down, and if it reaches
+-- zero wraps around to the initial value.
+wrappingCounter ::
+  (HiddenClockResetEnable dom, KnownNat n) =>
+  Unsigned n ->
+  Signal dom (Unsigned n)
+wrappingCounter upper = counter
+ where
+  counter = register upper (go <$> counter)
```
Agreed. I tried, but this is unfortunately more work than it seems due to the propagation of type variables throughout the rest of the code.

martijnbastiaan commented:
I've added more documentation and added an explanation on what we can improve (and why we didn't). I'm postponing merging so I can get simulation on CI.

martijnbastiaan commented:
I'm dumb. I promised myself to defer merging this to prevent merge conflicts with #161 :facepalm:
